sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -79,8 +91,145 @@ object GpuScalar {
+    //  `convertUTF8StringTo`. Tracked by https://github.com/rapidsai/cudf/issues/8137
+    case StringType => element.asInstanceOf[UTF8String] match {
+      case null => null
+      case s => s.toString
+    }
+    case dt: DecimalType =>
+      if (DecimalType.is32BitDecimalType(dt)) {
+        convertDecimalTo(element.asInstanceOf[Decimal], dt).intValue()
+      } else {
+        convertDecimalTo(element.asInstanceOf[Decimal], dt)
+      }
+    case ArrayType(eType, _) => element.asInstanceOf[ArrayData] match {
+      case null => null
+      case ar => ar.array.map(convertElementTo(_, eType)).toList.asJava
+    }
+    case StructType(fields) => element.asInstanceOf[InternalRow] match {
+      case null => null
+      case row =>
+        val data = fields.zipWithIndex.map { case (f, id) =>
+          convertElementTo(row.get(id, f.dataType), f.dataType)
+        }
+        new HostColumnVector.StructData(data.asInstanceOf[Array[Object]]: _*)
+    }
+    case _ => element
+  }
+
+  /**
+   * Creates a cudf Scalar from a ArrayData.
+   *
+   * It does not handle null because null is filtered out in 'from'.
+   *
+   * NOTE: This is done by leveraging the `ColumnVector.fromXXX` API family. These APIs are
+   * designed for test only according to its doc, but it is ok to use here, since the size
+   * of the literal values will not be large in most cases.
+   *
+   * @param data the data to build the Scalar. Should not be null.
+   * @param elementType the element type
+   * @return a cudf Scalar contains the array data.
+   */
+  private def createScalarFromArray(data: ArrayData, elementType: DataType): Scalar = {
+    val array = data.array
+    val listView = elementType match {
+      // Uses the boxed version for primitive types to keep the nulls.
+      case ByteType => ColumnVector.fromBoxedBytes(castTo[JByte](array): _*)
+      case DateType => ColumnVector.timestampDaysFromBoxedInts(castTo[Integer](array): _*)
+      case LongType => ColumnVector.fromBoxedLongs(castTo[JLong](array): _*)
+      case ShortType => ColumnVector.fromBoxedShorts(castTo[JShort](array): _*)
+      case FloatType => ColumnVector.fromBoxedFloats(castTo[JFloat](array): _*)
+      case DoubleType => ColumnVector.fromBoxedDoubles(castTo[JDouble](array): _*)
+      case IntegerType => ColumnVector.fromBoxedInts(castTo[Integer](array): _*)
+      case BooleanType => ColumnVector.fromBoxedBooleans(castTo[JBoolean](array): _*)
+      case TimestampType =>
+        ColumnVector.timestampMicroSecondsFromBoxedLongs(castTo[JLong](array): _*)
+      case StringType =>
+        // To be updated. https://github.com/rapidsai/cudf/issues/8137
+        val rows = castTo[UTF8String, Array[Byte]](array)(convertUTF8StringTo(_))
+        ColumnVector.build(DType.STRING, rows.length, b => {
+          rows.foreach {
+            case null => b.appendNull()
+            case s => b.appendUTF8String(s.asInstanceOf[Array[Byte]])
+          }
+        })
```

Comment:
Nit: If we're going to go through array elements one-by-one, to call `appendUTF8String`, why not skip the double-pass and extra array obj creation in `rows` and instead just walk `array `directly calling `appendUTF8String(e.getBytes)` for each element that is non-null.

Reply:
Good suggestion, will update later